Feature/add redis queue

yujinjcho opened the pull request:
For https://github.com/yujinjcho/movie_recommendations/issues/33.

I moved the recommendations calculation to a background job and added in polling to the ios api call. 

I noticed the polling stops if I change the ViewController from the RecommendationsTableViewController to the RateViewController. Do you know how I would keep it running?

yujinjcho committed:
Add redis background job / queuing

yujinjcho committed:
xcode updates

nyoon commented on ios_ui/MovieRec/RecommendationsTableViewController.swift:
```diff
@@ -7,6 +7,8 @@
 //
 
 import UIKit
+import Foundation
```
I don't think you need foundation here.

nyoon commented on ios_ui/MovieRec/RecommendationsTableViewController.swift:
```diff
@@ -66,23 +70,62 @@ class RecommendationsTableViewController: UITableViewController {
-            for recommendation in responseJSON["recommendations"]! {
-                self.recommendations += [Recommendation(title: recommendation)]
-            }
+    private func refreshMovies(data: JSON) -> Void {
```
Good use of SwiftyJSON

nyoon commented on ios_ui/MovieRec/RecommendationsTableViewController.swift:
```diff
@@ -66,23 +70,62 @@ class RecommendationsTableViewController: UITableViewController {
-            DispatchQueue.main.async {
-                self.tableView.reloadData()
-            }
+        DispatchQueue.main.async {
```
You probably want [unowned self] here so you dont run into retain cycles on this view controller

nyoon replied:
[unowned self] in

nyoon commented on ios_ui/MovieRec/RecommendationsTableViewController.swift:
```diff
@@ -66,23 +70,62 @@ class RecommendationsTableViewController: UITableViewController {
+    func startTimer(job_id: String) {
+        let queue = DispatchQueue(label: "com.domain.app.timer")
+        timer = DispatchSource.makeTimerSource(queue: queue)
+        timer!.scheduleRepeating(deadline: .now(), interval: .seconds(5))
```
I wouldn't force unwrap this timer object. I would use guard let timer = DispatchSource.makeTimerSource(queue: queue) else { fatalError("timer was not created") }

you can decide if this should fatal out or not, otherwise inside the else block you can just return. and maybe log something out indicating it messed up

nyoon commented on ios_ui/MovieRec/RecommendationsTableViewController.swift:
```diff
@@ -66,23 +70,62 @@ class RecommendationsTableViewController: UITableViewController {
-            }
+    private func refreshMovies(data: JSON) -> Void {
+        for recommendation in data.arrayValue {
+            self.recommendations += [Recommendation(title: recommendation.stringValue)]
```
also you don't need self here since it isn't inside a closure

nyoon commented on ios_ui/MovieRec/RecommendationsTableViewController.swift:
```diff
@@ -66,23 +70,62 @@ class RecommendationsTableViewController: UITableViewController {
+    
+    private func checkJobStatus(data: Data) -> Void {
+        let json = JSON(data: data)
+        if json["status"].stringValue == "completed" {
```
I would actually create an enum of some sort that indicates the status and use its rawValue as its type to be more type safe instead of using a raw string.

yujinjcho committed:
Update RecommendationsViewController

remove unnecessary reference to self

add enum for checking TrainingStatus

add capture list to avoid strong reference cycle

change force unwrap timer to guard let
